lte/gateway/c/oai/tasks/mme_app/MmeUeContext.cpp

```diff
@@ -0,0 +1,131 @@
+      msec, repeat, handler, id);
+}
+
+//------------------------------------------------------------------------------
+void mme_app_stop_timer(int timer_id) {
+  magma::lte::MmeUeContext::Instance().StopTimer(timer_id);
+}
+//------------------------------------------------------------------------------
+void mme_app_resume_timer(
+    struct ue_mm_context_s* const ue_mm_context_pP, time_t start_time,
+    struct mme_app_timer_t* timer, zloop_timer_fn timer_expiry_handler,
+    char* timer_name) {
+  OAILOG_FUNC_IN(LOG_MME_APP);
+  time_t current_time = time(NULL);
+  time_t lapsed_time  = current_time - start_time;
+  OAILOG_DEBUG(LOG_MME_APP, "Handling :%s timer \n", timer_name);
+
+  /* Below condition validates whether timer has expired before MME recovers
+   * from restart, so MME shall handle as timer expiry
+   */
+  if (timer->sec <= lapsed_time) {
+    timer_expiry_handler(mme_app_task_zmq_ctx.event_loop, timer->id, NULL);
+    OAILOG_FUNC_OUT(LOG_MME_APP);
+  }
+  uint32_t remaining_time_in_seconds = timer->sec - lapsed_time;
+  OAILOG_DEBUG(
+      LOG_MME_APP,
+      "Current_time :%ld %s timer start time :%ld "
+      "lapsed time:%ld remaining time:%d \n",
+      current_time, timer_name, start_time, lapsed_time,
+      remaining_time_in_seconds);
+
+  // Start timer only for remaining duration
+  if ((timer->id = magma::lte::MmeUeContext::Instance().StartTimer(
+           remaining_time_in_seconds * 1000, TIMER_REPEAT_ONCE,
+           timer_expiry_handler, ue_mm_context_pP->mme_ue_s1ap_id)) == -1) {
+    OAILOG_ERROR_UE(
+        LOG_MME_APP, ue_mm_context_pP->emm_context._imsi64,
+        "Failed to start %s timer for UE id "
+        "" MME_UE_S1AP_ID_FMT "\n",
+        timer_name, ue_mm_context_pP->mme_ue_s1ap_id);
+    timer->id = MME_APP_TIMER_INACTIVE_ID;
+  } else {
+    OAILOG_DEBUG_UE(
+        LOG_MME_APP, ue_mm_context_pP->emm_context._imsi64,
+        "Started %s timer for UE id " MME_UE_S1AP_ID_FMT "\n", timer_name,
+        ue_mm_context_pP->mme_ue_s1ap_id);
+  }
+  OAILOG_FUNC_OUT(LOG_MME_APP);
+}
+//------------------------------------------------------------------------------
+bool mme_app_get_timer_arg(int timer_id, timer_arg_t* arg) {
+  std::pair<magma::lte::TimerArgType, bool> p =
+      magma::lte::MmeUeContext::Instance().GetTimerArg(timer_id);
+  if (p.second) {
+    *arg = p.first;
+  }
+  return p.second;
+}
+
+namespace magma {
+namespace lte {
+//------------------------------------------------------------------------------
+int MmeUeContext::StartTimer(
+    size_t msec, timer_repeat_t repeat, zloop_timer_fn handler,
+    TimerArgType id) {
+  int timer_id = -1;
+  if ((timer_id = start_timer(
+           &mme_app_task_zmq_ctx, msec, repeat, handler, nullptr)) != -1) {
+    auto p = mme_app_timers.insert(std::pair<int, uint32_t>(timer_id, id));
+    if (not p.second) {
+      stop_timer(&mme_app_task_zmq_ctx, timer_id);
+      timer_id = -1;
+    }
+  }
```

Comment:
This looks a bit convoluted to me. Why start a timer in the first place if it exists already? UE context already stores the timer ids for a given UE. Let's for the sake of argument assume that there was an existing timer, why should we stop it?

Reply:
to be updated by @lionelgo .

Reply:
pushed a commit to @Vyas20